File: backend/src/utils/startup-check.js
/**
 * Development-mode startup diagnostics for matsya AI Backend.
 * Validates required environment variables and probes external service connectivity.
 * Only runs when NODE_ENV !== 'production'.
 */

const { DynamoDBClient, ListTablesCommand } = require('@aws-sdk/client-dynamodb');
const { S3Client, HeadBucketCommand } = require('@aws-sdk/client-s3');

// ── ANSI color helpers ──────────────────────────────────────────────────────
const c = {
    reset: '\x1b[0m',
    bold: '\x1b[1m',
    dim: '\x1b[2m',
    red: '\x1b[31m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    cyan: '\x1b[36m',
    white: '\x1b[37m',
    bgRed: '\x1b[41m',
    bgGreen: '\x1b[42m',
    bgYellow: '\x1b[43m',
};

const OK = `${c.green}✅${c.reset}`;
const WARN = `${c.yellow}⚠️${c.reset}`;
const FAIL = `${c.red}❌${c.reset}`;
const INFO = `${c.blue}ℹ️${c.reset}`;

// ── Check definitions ───────────────────────────────────────────────────────

const ENV_CHECKS = [
    // [name, envKey, level: 'critical'|'warn', description]
    ['S3_BUCKET_NAME', 'S3_BUCKET_NAME', 'critical', 'S3 bucket for image storage'],
    ['AWS_REGION', 'AWS_REGION', 'critical', 'AWS region for DynamoDB/S3'],
    ['COGNITO_USER_POOL_ID', 'COGNITO_USER_POOL_ID', 'critical', 'Cognito user pool for auth'],
    ['COGNITO_CLIENT_ID', 'COGNITO_CLIENT_ID', 'critical', 'Cognito client ID for auth'],
    ['DYNAMODB_IMAGES_TABLE', 'DYNAMODB_IMAGES_TABLE', 'warn', 'DynamoDB images table'],
    ['DYNAMODB_CHATS_TABLE', 'DYNAMODB_CHATS_TABLE', 'warn', 'DynamoDB chats table'],
    ['DYNAMODB_USERS_TABLE', 'DYNAMODB_USERS_TABLE', 'warn', 'DynamoDB users table'],
    ['GROUPS_TABLE', 'GROUPS_TABLE', 'warn', 'DynamoDB groups table'],
    ['ML_API_URL', 'ML_API_URL', 'warn', 'ML prediction API URL'],
    ['ML_API_KEY', 'ML_API_KEY', 'warn', 'ML API authentication key'],
    ['CHAT_API_URL', 'CHAT_API_URL', 'warn', 'Agent chat API URL'],
    ['OPENWEATHERMAP_API_KEY', 'OPENWEATHERMAP_API_KEY', 'warn', 'OpenWeatherMap API key'],
];

// ── Main diagnostics runner ─────────────────────────────────────────────────

async function runStartupChecks() {
    if (process.env.NODE_ENV === 'production') return { ok: true };

    const results = { criticalErrors: 0, warnings: 0, checks: [] };

    console.log('');
    console.log(`${c.cyan}${c.bold}╔══════════════════════════════════════════════════════════════════╗${c.reset}`);
    console.log(`${c.cyan}${c.bold}║  🐟 matsya AI Backend - Development Startup Diagnostics        ║${c.reset}`);
    console.log(`${c.cyan}${c.bold}╠══════════════════════════════════════════════════════════════════╣${c.reset}`);

    // ── 1. Environment Variables ─────────────────────────────────────────────
    console.log(`${c.cyan}${c.bold}║  ${c.white}Environment Variables${c.reset}`);
    console.log(`${c.cyan}${c.bold}╟──────────────────────────────────────────────────────────────────╢${c.reset}`);

    for (const [name, envKey, level, desc] of ENV_CHECKS) {
        const value = process.env[envKey];
        const maxLen = 30;

        if (value) {
            const display = value.length > maxLen ? value.slice(0, maxLen) + '…' : value;
            console.log(`${c.cyan}║${c.reset}  ${OK} ${name.padEnd(26)} = ${c.dim}${display}${c.reset}`);
            results.checks.push({ name, status: 'ok' });
        } else if (level === 'critical') {
            console.log(`${c.cyan}║${c.reset}  ${FAIL} ${name.padEnd(26)} = ${c.red}${c.bold}MISSING${c.reset}  ${c.dim}(${desc})${c.reset}`);
            results.criticalErrors++;
            results.checks.push({ name, status: 'critical' });
        } else {
            console.log(`${c.cyan}║${c.reset}  ${WARN} ${name.padEnd(26)} = ${c.yellow}not set${c.reset}  ${c.dim}(${desc})${c.reset}`);
            results.warnings++;
            results.checks.push({ name, status: 'warn' });
        }
    }

    // ── 2. Connectivity Probes ──────────────────────────────────────────────
    console.log(`${c.cyan}${c.bold}╟──────────────────────────────────────────────────────────────────╢${c.reset}`);
    console.log(`${c.cyan}${c.bold}║  ${c.white}Connectivity${c.reset}`);
    console.log(`${c.cyan}${c.bold}╟──────────────────────────────────────────────────────────────────╢${c.reset}`);

    // DynamoDB
    try {
        const ddb = new DynamoDBClient({ region: process.env.AWS_REGION || '' });
        const res = await ddb.send(new ListTablesCommand({}));
        const count = res.TableNames?.length || 0;
        console.log(`${c.cyan}║${c.reset}  ${OK} ${'DynamoDB'.padEnd(26)} = ${c.green}connected${c.reset} ${c.dim}(${count} tables)${c.reset}`);
    } catch (err) {
        console.log(`${c.cyan}║${c.reset}  ${FAIL} ${'DynamoDB'.padEnd(26)} = ${c.red}unreachable${c.reset}  ${c.dim}${err.message?.slice(0, 50)}${c.reset}`);
        results.criticalErrors++;
    }

    // S3 Bucket
    if (process.env.S3_BUCKET_NAME) {
        try {
            const s3 = new S3Client({ region: process.env.AWS_REGION || '' });
            await s3.send(new HeadBucketCommand({ Bucket: process.env.S3_BUCKET_NAME }));
            console.log(`${c.cyan}║${c.reset}  ${OK} ${'S3 Bucket'.padEnd(26)} = ${c.green}accessible${c.reset}`);
        } catch (err) {
            console.log(`${c.cyan}║${c.reset}  ${WARN} ${'S3 Bucket'.padEnd(26)} = ${c.yellow}error${c.reset}  ${c.dim}${err.message?.slice(0, 50)}${c.reset}`);
            results.warnings++;
        }
    }

    // Agent API
    const agentUrl = process.env.CHAT_API_URL?.replace('/chat', '') || '';
    try {
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), 3000);
        const res = await fetch(`${agentUrl}/health`, { signal: controller.signal });
        clearTimeout(timeout);
        if (res.ok) {
            console.log(`${c.cyan}║${c.reset}  ${OK} ${'Agent API'.padEnd(26)} = ${c.green}${agentUrl}${c.reset} ${c.dim}(healthy)${c.reset}`);
        } else {
            console.log(`${c.cyan}║${c.reset}  ${WARN} ${'Agent API'.padEnd(26)} = ${c.yellow}${agentUrl}${c.reset} ${c.dim}(status ${res.status})${c.reset}`);
            results.warnings++;
        }
    } catch {
        console.log(`${c.cyan}║${c.reset}  ${WARN} ${'Agent API'.padEnd(26)} = ${c.yellow}unreachable${c.reset}  ${c.dim}${agentUrl}${c.reset}`);
        results.warnings++;
    }

    // ML API
    const mlUrl = process.env.ML_API_URL || '';
    if (mlUrl) {
        try {
            const controller = new AbortController();
            const timeout = setTimeout(() => controller.abort(), 5000);
            const res = await fetch(mlUrl.replace('/predict', '/'), { signal: controller.signal });
            clearTimeout(timeout);
            console.log(`${c.cyan}║${c.reset}  ${OK} ${'ML API'.padEnd(26)} = ${c.green}reachable${c.reset} ${c.dim}(status ${res.status})${c.reset}`);
        } catch {
            console.log(`${c.cyan}║${c.reset}  ${WARN} ${'ML API'.padEnd(26)} = ${c.yellow}unreachable${c.reset}  ${c.dim}${mlUrl}${c.reset}`);
            results.warnings++;
        }
    }

    // ── 3. Summary ──────────────────────────────────────────────────────────
    console.log(`${c.cyan}${c.bold}╟──────────────────────────────────────────────────────────────────╢${c.reset}`);

    if (results.criticalErrors > 0) {
        console.log(`${c.cyan}║${c.reset}  ${c.bgRed}${c.white}${c.bold} RESULT ${c.reset} ${c.red}${results.criticalErrors} critical error(s)${c.reset}, ${c.yellow}${results.warnings} warning(s)${c.reset}`);
    } else if (results.warnings > 0) {
        console.log(`${c.cyan}║${c.reset}  ${c.bgYellow}${c.bold} RESULT ${c.reset} ${c.green}0 critical errors${c.reset}, ${c.yellow}${results.warnings} warning(s)${c.reset}`);
    } else {
        console.log(`${c.cyan}║${c.reset}  ${c.bgGreen}${c.bold} RESULT ${c.reset} ${c.green}All checks passed!${c.reset}`);
    }

    console.log(`${c.cyan}${c.bold}╚══════════════════════════════════════════════════════════════════╝${c.reset}`);
    console.log('');

    return { ok: results.criticalErrors === 0, ...results };
}

module.exports = { runStartupChecks };

File: working/agent/src/utils/startup_check.py
"""
Development-mode startup diagnostics for matsya AI Agent.
Validates required environment variables and probes external service connectivity.
Only runs when APP_ENV != 'production'.
"""
from __future__ import annotations
import os
import logging

logger = logging.getLogger(__name__)

# ── ANSI color helpers ──────────────────────────────────────────────────────
class C:
    RESET   = '\033[0m'
    BOLD    = '\033[1m'
    DIM     = '\033[2m'
    RED     = '\033[31m'
    GREEN   = '\033[32m'
    YELLOW  = '\033[33m'
    BLUE    = '\033[34m'
    CYAN    = '\033[36m'
    WHITE   = '\033[37m'
    BG_RED  = '\033[41m'
    BG_GREEN= '\033[42m'
    BG_YELLOW='\033[43m'

OK   = f'{C.GREEN}✅{C.RESET}'
WARN = f'{C.YELLOW}⚠️{C.RESET}'
FAIL = f'{C.RED}❌{C.RESET}'

# ── Check definitions ───────────────────────────────────────────────────────
ENV_CHECKS = [
    # (name, env_key, level, description)
    ('GOOGLE_API_KEY',              'GOOGLE_API_KEY',               'critical', 'Gemini API key for LLM'),
    ('COGNITO_USER_POOL_ID',        'COGNITO_USER_POOL_ID',         'critical', 'Cognito User Pool for auth'),
    ('COGNITO_CLIENT_ID',           'COGNITO_CLIENT_ID',            'warn',     'Cognito Client ID'),
    ('DB_HOST',                     'DB_HOST',                      'critical', 'MySQL host'),
    ('DB_USER',                     'DB_USER',                      'critical', 'MySQL user'),
    ('DB_PASSWORD',                 'DB_PASSWORD',                  'warn',     'MySQL password'),
    ('DB_NAME',                     'DB_NAME',                      'critical', 'MySQL database name'),
    ('OPENWEATHERMAP_API_KEY',      'OPENWEATHERMAP_API_KEY',       'warn',     'Weather API key'),
]


async def run_startup_checks() -> dict:
    """Run all startup diagnostics. Returns dict with 'ok' bool and counts."""
    if os.getenv('APP_ENV', 'development') == 'production':
        return {'ok': True}

    results = {'critical_errors': 0, 'warnings': 0}
    lines: list[str] = []

    lines.append('')
    lines.append(f'{C.CYAN}{C.BOLD}╔══════════════════════════════════════════════════════════════════╗{C.RESET}')
    lines.append(f'{C.CYAN}{C.BOLD}║  🤖 matsya AI Agent - Development Startup Diagnostics          ║{C.RESET}')
    lines.append(f'{C.CYAN}{C.BOLD}╠══════════════════════════════════════════════════════════════════╣{C.RESET}')

    # ── 1. Environment Variables ─────────────────────────────────────────────
    lines.append(f'{C.CYAN}{C.BOLD}║  {C.WHITE}Environment Variables{C.RESET}')
    lines.append(f'{C.CYAN}{C.BOLD}╟──────────────────────────────────────────────────────────────────╢{C.RESET}')

    for name, env_key, level, desc in ENV_CHECKS:
        value = os.getenv(env_key, '')
        max_len = 30

        if value:
            display = value[:max_len] + '…' if len(value) > max_len else value
            lines.append(f'{C.CYAN}║{C.RESET}  {OK} {name:<30} = {C.DIM}{display}{C.RESET}')
        elif level == 'critical':
            lines.append(f'{C.CYAN}║{C.RESET}  {FAIL} {name:<30} = {C.RED}{C.BOLD}MISSING{C.RESET}  {C.DIM}({desc}){C.RESET}')
            results['critical_errors'] += 1
        else:
            lines.append(f'{C.CYAN}║{C.RESET}  {WARN} {name:<30} = {C.YELLOW}not set{C.RESET}  {C.DIM}({desc}){C.RESET}')
            results['warnings'] += 1

    # ── 2. Connectivity Probes ──────────────────────────────────────────────
    lines.append(f'{C.CYAN}{C.BOLD}╟──────────────────────────────────────────────────────────────────╢{C.RESET}')
    lines.append(f'{C.CYAN}{C.BOLD}║  {C.WHITE}Connectivity{C.RESET}')
    lines.append(f'{C.CYAN}{C.BOLD}╟──────────────────────────────────────────────────────────────────╢{C.RESET}')

    # MySQL
    try:
        import pymysql
        ssl_enabled = os.getenv('DB_SSL', 'false').lower() == 'true'
        conn = pymysql.connect(
            host=os.getenv('DB_HOST', 'localhost'),
            port=int(os.getenv('DB_PORT', '3306')),
            user=os.getenv('DB_USER', 'root'),
            password=os.getenv('DB_PASSWORD', ''),
            database=os.getenv('DB_NAME', 'defaultdb'),
            charset='utf8mb4',
            connect_timeout=5,
            **({"ssl": {}} if ssl_enabled else {}),
        )
        with conn.cursor() as cur:
            cur.execute('SHOW TABLES')
            count = len(cur.fetchall())
        conn.close()
        lines.append(f'{C.CYAN}║{C.RESET}  {OK} {"MySQL":<30} = {C.GREEN}connected{C.RESET} {C.DIM}({count} tables){C.RESET}')
    except Exception as e:
        lines.append(f'{C.CYAN}║{C.RESET}  {FAIL} {"MySQL":<30} = {C.RED}unreachable{C.RESET}  {C.DIM}{str(e)[:50]}{C.RESET}')
        results['critical_errors'] += 1

    # Gemini API - lightweight HTTP check via REST endpoint
    api_key = os.getenv('GOOGLE_API_KEY', '')
    if api_key:
        try:
            import urllib.request
            import urllib.error
            model_name = os.getenv('GEMINI_MODEL', 'gemini-2.0-flash')
            url = f"https://generativelanguage.googleapis.com/v1beta/models/{model_name}?key={api_key}"
            req = urllib.request.Request(url, method='GET')
            req.add_header('Content-Type', 'application/json')
            with urllib.request.urlopen(req, timeout=5) as resp:
                if resp.status == 200:
                    lines.append(f'{C.CYAN}║{C.RESET}  {OK} {"Gemini API":<30} = {C.GREEN}connected{C.RESET} {C.DIM}({model_name}){C.RESET}')
                else:
                    lines.append(f'{C.CYAN}║{C.RESET}  {WARN} {"Gemini API":<30} = {C.YELLOW}status {resp.status}{C.RESET}')
                    results['warnings'] += 1
        except urllib.error.HTTPError as e:
            lines.append(f'{C.CYAN}║{C.RESET}  {FAIL} {"Gemini API":<30} = {C.RED}HTTP {e.code}{C.RESET}  {C.DIM}{str(e.reason)[:40]}{C.RESET}')
            results['critical_errors'] += 1
        except Exception as e:
            lines.append(f'{C.CYAN}║{C.RESET}  {FAIL} {"Gemini API":<30} = {C.RED}error{C.RESET}  {C.DIM}{str(e)[:50]}{C.RESET}')
            results['critical_errors'] += 1
    else:
        lines.append(f'{C.CYAN}║{C.RESET}  {FAIL} {"Gemini API":<30} = {C.RED}no API key{C.RESET}')
        results['critical_errors'] += 1

    # ── 3. Summary ──────────────────────────────────────────────────────────
    lines.append(f'{C.CYAN}{C.BOLD}╟──────────────────────────────────────────────────────────────────╢{C.RESET}')

    if results['critical_errors'] > 0:
        lines.append(f'{C.CYAN}║{C.RESET}  {C.BG_RED}{C.WHITE}{C.BOLD} RESULT {C.RESET} {C.RED}{results["critical_errors"]} critical error(s){C.RESET}, {C.YELLOW}{results["warnings"]} warning(s){C.RESET}')
    elif results['warnings'] > 0:
        lines.append(f'{C.CYAN}║{C.RESET}  {C.BG_YELLOW}{C.BOLD} RESULT {C.RESET} {C.GREEN}0 critical errors{C.RESET}, {C.YELLOW}{results["warnings"]} warning(s){C.RESET}')
    else:
        lines.append(f'{C.CYAN}║{C.RESET}  {C.BG_GREEN}{C.BOLD} RESULT {C.RESET} {C.GREEN}All checks passed!{C.RESET}')

    lines.append(f'{C.CYAN}{C.BOLD}╚══════════════════════════════════════════════════════════════════╝{C.RESET}')
    lines.append('')

    print('\n'.join(lines))

    results['ok'] = results['critical_errors'] == 0
    return results
